thorlcr/activities/external/thexternalslave.cpp

```diff
@@ -0,0 +1,175 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2018 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "jlib.hpp"
+#include "jset.hpp"
+#include "jqueue.tpp"
+#include "commonext.hpp"
+
+#include "thormisc.hpp"
+#include "thexception.hpp"
+#include "thbufdef.hpp"
+#include "thalloc.hpp"
+#include "eclrtl_imp.hpp"
+
+#include "slave.ipp"
+#include "thactivityutil.ipp"
+
+/////
+///////////////////
+//
+// ExternalSlaveActivity
+//
+
+class CNodeActivityContext : public IThorActivityContext
+{
+public:
+    CNodeActivityContext(bool _local, unsigned _numSlaves, unsigned _curSlave)
+    : local(_local), slaves(_local ? 1 : _numSlaves), curSlave(_local ? 0 : _curSlave)
+    {
+        assertex(curSlave < slaves);
+    }
+
+    virtual bool isLocal() const override { return local; }
+    virtual unsigned numSlaves() const override { return slaves; }
+    virtual unsigned numStrands() const override { return 1; }
+    virtual unsigned querySlave() const override { return curSlave; }
+    virtual unsigned queryStrand() const override { return 0; }
+protected:
+    unsigned slaves;
+    unsigned curSlave;
+    bool local;
+};
+
+class ExternalSlaveActivity : public CSlaveActivity
```

Comment:
picky: you changed NodeActivityContext to CNodeActivityContext, would be good to change this and ExternalSinkSlaveActivity further down.